petrock/vision.py

```diff
@@ -1,61 +1,49 @@
 from PIL import Image
-
 import cv2
+from petrock.llms import summon_moondream
+from openai import OpenAI
 import io
 import base64
 import logging
 
-from petrock.llms import summon_moondream
-
-
 class Webcam:
-    def get_image(self) -> Image:
+    def get_image(self, save_path=None) -> Image:
         raise NotImplementedError("Webcam class must implement `get_image()` method.")
 
-#capture image from cam, using openCV and return as PIL Image.
 class OpenCVWebcam(Webcam):
-    def get_image(self) -> Image:
-        cap = cv2.VideoCapture(0)  # 0 is usually the default camera
+    def get_image(self, save_path=None) -> Image:
+        cap = cv2.VideoCapture(0)  # Use the first webcam
         success, frame = cap.read()
         cap.release()
         if not success:
             raise Exception("Failed to capture image")
-        # Convert the color from BGR to RGB
         frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
-        return Image.fromarray(frame)
-
+        image = Image.fromarray(frame)
+        if save_path:
+            image.save(save_path)
+        print("Using standard webcam via OpenCV.")
+        return image
 
-# Function to encode the image directly from a PIL Image object
 def encode_image_to_base64(img: Image) -> str:
     buffered = io.BytesIO()
     img.save(buffered, format="JPEG")
     return base64.b64encode(buffered.getvalue()).decode('utf-8')
 
-# Function to encode the image
-def encode_local_image(image_path) -> str:
-    with open(image_path, "rb") as image_file:
-        return base64.b64encode(image_file.read()).decode('utf-8')
-
+def summon_moondream():
+    # Returns an initialized OpenAI object configured to communicate with your server
+    return OpenAI(api_key='moondream', base_url='http://localhost:8080/v1')
 
 class Vision:
-    def __init__(self, webcam: Webcam=None):
-       self.webcam = webcam
+    def __init__(self):
+        self.webcam = OpenCVWebcam()
 
-    def caption_image(self, img: Image) -> str:
-        base64_image = encode_image_to_base64(img)
+    def get_caption(self, save_path="./tmp/captured_image.jpg"):
```

Comment:
Refactor this to use the existing functions, and instead of just calling `vision.get_caption()` call `vision.webcam.get_image()` followed by whichever captioning system you want.

Captioning is logically distinct from using the webcam and should be usable independently.